File: pulsaros-spotlight-launcher/src/pulsaros_spotlight/clipboard.py
"""Clipboard history manager and auto-paste provider for PulsarOS Spotlight."""

from __future__ import annotations

import json
import logging
import os
import shutil
import subprocess
import time
from pathlib import Path
from typing import TYPE_CHECKING

import gi

gi.require_version("Gdk", "4.0")
gi.require_version("Gtk", "4.0")
from gi.repository import Gdk, GLib, Gtk

from pulsaros_spotlight.config import Config

if TYPE_CHECKING:
    from pulsaros_spotlight.search import SearchResult

logger = logging.getLogger(__name__)

DATA_DIR = Path(GLib.get_user_data_dir()) / "pulsaros-spotlight"
CLIPBOARD_FILE = DATA_DIR / "clipboard_history.json"


class ClipboardManager:
    """Monitors clipboard changes, stores history, and simulates auto-pasting."""

    def __init__(self, config: Config | None = None) -> None:
        self._config = config or Config()
        self._items: list[dict] = []
        self._load()
        self._init_listener()

    def _load(self) -> None:
        if CLIPBOARD_FILE.exists():
            try:
                self._items = json.loads(CLIPBOARD_FILE.read_text(encoding="utf-8"))
            except (json.JSONDecodeError, OSError):
                self._items = []

    def _save(self) -> None:
        DATA_DIR.mkdir(parents=True, exist_ok=True)
        try:
            CLIPBOARD_FILE.write_text(json.dumps(self._items, indent=2), encoding="utf-8")
        except OSError:
            pass

    def _init_listener(self) -> None:
        try:
            display = Gdk.Display.get_default()
            if display:
                clipboard = display.get_clipboard()
                clipboard.connect("changed", self._on_clipboard_changed)
        except Exception:
            logger.warning("Failed to hook Gdk.Clipboard listener", exc_info=True)

    def _on_clipboard_changed(self, clipboard: Gdk.Clipboard) -> None:
        try:
            clipboard.read_text_async(None, self._on_text_read)
        except Exception:
            pass

    def _on_text_read(self, clipboard: Gdk.Clipboard, result) -> None:
        try:
            text = clipboard.read_text_finish(result)
            if text and text.strip():
                self.record_clip(text.strip())
        except Exception:
            pass

    def record_clip(self, text: str) -> None:
        """Add a text snippet to clipboard history."""
        if not text:
            return
        clean = text.strip()
        if not clean:
            return

        # Avoid duplicate at top
        if self._items and self._items[0].get("text") == clean:
            return

        # Remove previous occurrence if exists
        self._items = [item for item in self._items if item.get("text") != clean]

        # Insert at front
        self._items.insert(
            0,
            {
                "text": clean,
                "timestamp": time.time(),
            },
        )

        # Enforce max items
        max_items = getattr(self._config, "clipboard_max_items", 50)
        self._items = self._items[:max_items]
        self._save()

    def search(self, query: str = "") -> list[SearchResult]:
        """Search clipboard history entries."""
        from pulsaros_spotlight.search import SearchResult

        # Also check current clipboard content on search
        try:
            p = subprocess.run(["wl-paste", "--no-newline"], capture_output=True, text=True, timeout=0.2)
            if p.returncode == 0 and p.stdout and p.stdout.strip():
                self.record_clip(p.stdout.strip())
        except Exception:
            pass

        results: list[SearchResult] = []
        q_lower = query.lower().strip()

        for idx, item in enumerate(self._items):
            text = item.get("text", "")
            if not text:
                continue

            if not q_lower or q_lower in text.lower():
                lines = [line.strip() for line in text.splitlines() if line.strip()]
                first_line = lines[0] if lines else text
                if len(first_line) > 60:
                    first_line = first_line[:57] + "..."

                preview = text.replace("\n", " ⏎ ")
                if len(preview) > 100:
                    preview = preview[:97] + "..."

                results.append(
                    SearchResult(
                        url=f"clipboard://{idx}",
                        title=first_line,
                        mime="text/plain-clipboard",
                        snippet=preview,
                        app=None,
                    )
                )

        return results

    def get_clip_by_index(self, index: int) -> str | None:
        if 0 <= index < len(self._items):
            return self._items[index].get("text")
        return None

    def paste_clip(self, text: str) -> None:
        """Set clipboard text and simulate Ctrl+V keystroke into previous window."""
        if not text:
            return

        # 1. Set to Gdk Clipboard and wl-copy
        try:
            display = Gdk.Display.get_default()
            if display:
                display.get_clipboard().set(text)
        except Exception:
            pass

        try:
            p = subprocess.Popen(["wl-copy"], stdin=subprocess.PIPE)
            p.communicate(text.encode("utf-8"))
        except Exception:
            pass

        # 2. Simulate Paste (Ctrl+V) after window hides and focus returns
        if getattr(self._config, "clipboard_auto_paste", True):
            GLib.timeout_add(100, self._simulate_paste)

    @staticmethod
    def _simulate_paste() -> bool:
        """Simulate Ctrl+V keystroke via ydotool, GNOME Shell extension D-Bus, or xdotool."""
        # 1. Native Wayland uinput via ydotool
        if shutil.which("ydotool"):
            try:
                env = os.environ.copy()
                sock = f"/run/user/{os.getuid()}/.ydotool_socket"
                if os.path.exists(sock):
                    env["YDOTOOL_SOCKET"] = sock
                subprocess.Popen(["ydotool", "key", "29:1", "47:1", "47:0", "29:0"], env=env)
                return False
            except Exception:
                pass

        # 2. Native Wayland GNOME Shell Extension D-Bus key injection
        try:
            bus = Gio.bus_get_sync(Gio.BusType.SESSION, None)
            proxy = Gio.DBusProxy.new_sync(
                bus,
                Gio.DBusProxyFlags.DO_NOT_LOAD_PROPERTIES,
                None,
                "org.gnome.Shell.Extensions.PulsarSpotlight",
                "/org/gnome/Shell/Extensions/PulsarSpotlight",
                "org.gnome.Shell.Extensions.PulsarSpotlight",
                None,
            )
            proxy.call_sync("Paste", None, Gio.DBusCallFlags.NONE, 500, None)
            return False
        except Exception:
            pass

        # 3. External tool fallback
        if shutil.which("xdotool"):
            try:
                subprocess.Popen(["xdotool", "key", "ctrl+v"])
                return False
            except Exception:
                pass

        return False

File: pulsaros-spotlight-launcher/src/clipboard.rs
use crate::config::SpotlightConfig;
use crate::search::SearchResult;
use gtk4::gdk;
use gtk4::prelude::*;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::sync::{Arc, Mutex};
use std::time::SystemTime;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ClipboardItem {
    pub text: String,
    pub timestamp: f64,
}

#[derive(Clone)]
pub struct ClipboardManager {
    inner: Arc<Mutex<ClipboardManagerInner>>,
}

struct ClipboardManagerInner {
    items: Vec<ClipboardItem>,
    config: SpotlightConfig,
}

impl ClipboardManager {
    fn data_dir() -> PathBuf {
        dirs::data_dir()
            .unwrap_or_else(|| PathBuf::from("/home/jaime/.local/share"))
            .join("pulsaros-spotlight")
    }

    fn clipboard_file() -> PathBuf {
        Self::data_dir().join("clipboard_history.json")
    }

    pub fn new(config: SpotlightConfig) -> Self {
        let mut items = Vec::new();
        let file = Self::clipboard_file();
        if file.exists() {
            if let Ok(content) = fs::read_to_string(&file) {
                if let Ok(loaded) = serde_json::from_str::<Vec<ClipboardItem>>(&content) {
                    items = loaded;
                }
            }
        }

        let inner = Arc::new(Mutex::new(ClipboardManagerInner { items, config }));
        Self { inner }
    }

    pub fn save(&self) {
        let inner = self.inner.lock().unwrap();
        let dir = Self::data_dir();
        let _ = fs::create_dir_all(&dir);
        let file = Self::clipboard_file();
        if let Ok(content) = serde_json::to_string_pretty(&inner.items) {
            let _ = fs::write(&file, content);
        }
    }

    pub fn record_clip(&self, text: &str) {
        let clean = text.trim();
        if clean.is_empty() {
            return;
        }

        let mut inner = self.inner.lock().unwrap();
        // Avoid duplicate at top
        if let Some(first) = inner.items.first() {
            if first.text == clean {
                return;
            }
        }

        // Remove previous occurrences
        inner.items.retain(|item| item.text != clean);

        let timestamp = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);

        inner.items.insert(0, ClipboardItem {
            text: clean.to_string(),
            timestamp,
        });

        // Enforce max items
        let max_items = inner.config.clipboard_max_items;
        if inner.items.len() > max_items {
            inner.items.truncate(max_items);
        }

        // Save immediately
        drop(inner);
        self.save();
    }

    pub fn get_items(&self) -> Vec<ClipboardItem> {
        self.inner.lock().unwrap().items.clone()
    }

    pub fn get_clip_by_index(&self, index: usize) -> Option<String> {
        let inner = self.inner.lock().unwrap();
        inner.items.get(index).map(|item| item.text.clone())
    }

    pub fn search_history(&self, query: &str) -> Vec<SearchResult> {
        let items = self.get_items();
        let q_lower = query.trim().to_lowercase();
        let mut results = Vec::new();

        for (idx, item) in items.iter().enumerate() {
            if q_lower.is_empty() || item.text.to_lowercase().contains(&q_lower) {
                let lines: Vec<&str> = item.text.lines().map(|l| l.trim()).filter(|l| !l.is_empty()).collect();
                let mut first_line = lines.first().copied().unwrap_or(&item.text).to_string();
                if first_line.chars().count() > 60 {
                    first_line = first_line.chars().take(57).collect::<String>() + "...";
                }

                let mut preview = item.text.replace('\n', " ⏎ ");
                if preview.chars().count() > 100 {
                    preview = preview.chars().take(97).collect::<String>() + "...";
                }

                results.push(SearchResult {
                    url: format!("clipboard://{}", idx),
                    title: first_line,
                    mime: "text/plain-clipboard".to_string(),
                    snippet: preview,
                    app: None,
                });
            }
        }

        results
    }

    pub fn paste_clip(&self, text: &str) {
        if text.is_empty() {
            return;
        }

        // 1. Set to Gdk Clipboard
        if let Some(display) = gdk::Display::default() {
            display.clipboard().set_text(text);
        }

        // 2. Run wl-copy in background
        if let Ok(mut child) = Command::new("wl-copy")
            .stdin(Stdio::piped())
            .spawn()
        {
            if let Some(mut stdin) = child.stdin.take() {
                use std::io::Write;
                let _ = stdin.write_all(text.as_bytes());
            }
        }

        // 3. Simulate Paste (Ctrl+V) after window hides
        let auto_paste = self.inner.lock().unwrap().config.clipboard_auto_paste;
        if auto_paste {
            gtk4::glib::timeout_add_local(std::time::Duration::from_millis(100), move || {
                Self::simulate_paste();
                gtk4::glib::ControlFlow::Break
            });
        }
    }

    fn simulate_paste() {
        let which = |cmd: &str| -> Option<PathBuf> {
            std::env::var_os("PATH").and_then(|paths| {
                std::env::split_paths(&paths).filter_map(|dir| {
                    let full_path = dir.join(cmd);
                    if full_path.is_file() {
                        Some(full_path)
                    } else {
                        None
                    }
                }).next()
            })
        };

        // 1. Native Wayland uinput via ydotool
        if let Some(path) = which("ydotool") {
            let uid = unsafe { libc::getuid() };
            let sock = format!("/run/user/{}/.ydotool_socket", uid);
            let mut cmd = Command::new(path);
            cmd.args(&["key", "29:1", "47:1", "47:0", "29:0"]);
            if std::path::Path::new(&sock).exists() {
                cmd.env("YDOTOOL_SOCKET", sock);
            }
            if cmd.spawn().is_ok() {
                return;
            }
        }

        // 2. GNOME Shell Extension D-Bus key injection
        if let Ok(conn) = zbus::blocking::Connection::session() {
            let body = zbus::zvariant::Value::from(""); // empty or dummy
            // Make session call
            if conn.call_method(
                Some("org.gnome.Shell.Extensions.PulsarSpotlight"),
                "/org/gnome/Shell/Extensions/PulsarSpotlight",
                Some("org.gnome.Shell.Extensions.PulsarSpotlight"),
                "Paste",
                &body,
            ).is_ok() {
                return;
            }
        }

        // 3. Fallback: xdotool
        if let Some(path) = which("xdotool") {
            let _ = Command::new(path)
                .args(&["key", "ctrl+v"])
                .spawn();
        }
    }
}
